File: advent_of_code_2021/2021_day_06_part_2.py
import re, sys
with open(sys.argv[1]) as f:
        lines = [ re.findall( r'\d+', line ) for line in f ]
timer = {}
for t in map( int, lines[0] ):
    timer[t] = timer.get(t,0) + 1
for day in range(256):
    timer_next = {}
    for t in timer:
        n = 7 if t == 8 else (t-1) % 7
        timer_next[n] = timer_next.get(n,0) + timer[t]
    timer_next[8] = timer.get(0,0)
    timer = timer_next
    if   day == 80-1:
        print("Answer part 1: ", sum(timer.values()))
    elif day == 256-1:
        print("Answer part 2: ", sum(timer.values()))

File: advent_of_code_2021/2021_day_06_part_2.pl
use v5.10;
my %timer; $timer{$_}++ for <> =~ /\d+/g;
for ( 1 .. 256 ) {
    my %timer_next;
    for( keys %timer ){
        $timer_next{ $_ == 8 ? 7 : ($_-1) % 7 } += $timer{$_}
    }
    %timer = ( %timer_next, 8 => $timer{0} );
}
say "Answer: " . eval join '+', values %timer;
